pkg/sentry/fsimpl/gofer/filesystem.go

```diff
@@ -407,7 +407,16 @@ func (fs *filesystem) doCreateAt(ctx context.Context, rp *vfs.ResolvingPath, dir
 	if err != nil {
 		return err
 	}
-	if err := parent.checkPermissions(rp.Credentials(), vfs.MayWrite|vfs.MayExec); err != nil {
+
+	parent.dirMu.Lock()
+	defer parent.dirMu.Unlock()
+
+	// Check for existence before checking for permissions, like Linux does.
+	_, err = fs.stepLocked(ctx, rp, parent, true /* mayFollowSymlinks */, &ds)
+	if err == nil {
+		return syserror.EEXIST
```

Comment:
I think we still need to check Exec permission against the parent to verify that the caller can list the files in the folder.

```
mkdir -p /noperm/exists
chmod 766 /noperm
sudo -u nobody mkdir /noperm/exists
mkdir: cannot create directory ‘/noperm/exists’: Permission denied
```

Reply:
Thanks for the review. The `exec` check is done as part of `stepLocked()`. As you can see in the test case (mkdir.cc) there is a test for `0600` that ensure `mkdir` returns `EACCES`. I'll add a comment...

Reply:
Ahh sorry for not following that further.  Great!  LGTM